client/question/services/FeedbackGeneratorService.js

```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
     };
 
     /**
+     * Concats two strings to create testCaseKey
+     *
+     * @param {string} testSuiteId
+     * @param {string} testCaseIndex
+     */
+    var _getTestCaseKey = function(testSuiteId, testCaseIndex) {
+      return testSuiteId + '-' + testCaseIndex;
+    };
+
+    /**
      * Returns the Feedback object created as a result of a specified
      * correctness test.
      *
-     * @param {string|null} outputFunctionName If needed, is string of the
-     *    name for the function needed to process/format the user's output.
      * @param {CorrectnessTest} TestCase
+     * @param {number} testCaseIndex
      * @param {*} observedOutput Actual output for running user's code.
+     * @param {string} testSuiteId
      * @returns {Feedback}
      * @private
      */
     var _getCorrectnessTestFeedback = function(
-      outputFunctionName, testCase, observedOutput) {
+      testCase, testCaseIndex, observedOutput, testSuiteId) {
+      var testCaseKey = _getTestCaseKey(testSuiteId, testCaseIndex);
       var allowedOutputExample = testCase.getAnyAllowedOutput();
       var feedback = FeedbackObjectFactory.create(
         FEEDBACK_CATEGORIES.INCORRECT_OUTPUT_FAILURE);
-      // TODO(eyurko): Add varied statements for when code is incorrect.
-      feedback.appendTextParagraph('Your code produced the following result:');
-      if (outputFunctionName) {
+      // Check if new / next test suite
+      if (testSuiteId !== previousTestSuiteId) {
+        previousTestSuiteId = testSuiteId;
+        // Catch regressions
```

Comment:
Copying comment from previous review:

> Just to check, is the logic here correct? For example, if the user has passed suite 1 and 2 without getting any feedback, and subsequently regresses, would there even be an entry for those test cases in correctnessTestStates?

Reply:
It should not. IIUC, if they get no feedback at all, then they would also get no correctness feedback. In that case, there should be no entry in correctnessTestStates. If they fail a test later, it would be as if it was the first time (not a regression); at that point, an entry is made in correctnessTestStates. Since testCaseKey includes both the suite ID and test case index, it should hold across different test suites.